scripts/zones/AlTaieu/mobs/Absolute_Virtue.lua

```diff
@@ -4,20 +4,274 @@
+local debug   = utils.getDebugPlayerPrinter(debugAV)
+
 local entity = {}
 
+entity.handleDamageResists = function(mob)
+    local next = mob:getLocalVar('dmgThreshold')
+    local current = mob:getHPP()
+    if current <= next then
+        mob:setLocalVar('dmgThreshold', next - 10)
+        local dmg = (100 - current) * -1
+        mob:setMod(xi.mod.UDMGPHYS, dmg)
+        mob:setMod(xi.mod.UDMGRANGE, dmg)
+        mob:setMod(xi.mod.UDMGMAGIC, dmg * 10)
+        mob:setMod(xi.mod.UDMGBREATH, dmg)
+    end
+end
+
+local playerAbilityToMobSP =
+{
+    [16] = 688, -- MIGHTY_STRIKES
+    [17] = 690, -- HUNDRED_FISTS
+    [18] = 689, -- BENEDICTION
+    [19] = 691, -- MANAFONT
+    [20] = 692, -- CHAINSPELL
+    [21] = 693, -- PERFECT_DODGE
+    [22] = 694, -- INVINCIBLE
+    [23] = 695, -- BLOOD_WEAPON
+    [24] = 740, -- FAMILIAR
+    [25] = 696, -- SOUL_VOICE
+    [26] = 1389, -- EES_AERN
+    [27] = 730, -- MEIKYO_SHISUI
+    [28] = 731, -- MIJIN_GAKURE
+    [30] = 734, -- ASTRAL_FLOW
+    [61] = 732, -- CALL_WYVERN
+}
+
+entity.isLocked = function(sp)
+    if #xi.av.sps == 0 then
+        return true
+    end
+
+    for _, jsa in ipairs(xi.av.sps) do
+        if sp == jsa then
+            return false
+        end
+    end
+
+    return true
+end
+
+entity.lock = function(sp)
+    for i, jsa in ipairs(xi.av.sps) do
+        if jsa == sp then
+            debug(string.format('locked: %d', sp))
+            table.remove(xi.av.sps, i)
+            break
+        end
+    end
+
+    for i, jsa in ipairs(xi.av.braceletsps) do
+        if jsa == sp then
+            table.remove(xi.av.braceletsps, i)
+            break
+        end
+    end
+end
+
+-- TODO: handle pets and pet sp abilities
+entity.handleSP = function(mob)
+    local now = os.time()
+    if now > xi.av.nextsp then
+        if xi.av.bracelets and #xi.av.braceletsps ~= 0 then
+            local trigger = xi.av.braceletsps[math.random(1, #xi.av.braceletsps)]
+            local combo = combos[trigger]
+            if trigger == xi.jsa.CHAINSPELL then
+                combo = combos[math.random(1, 2)]
+            end
+
+            for _, jsa in ipairs(combo) do
+                if not entity.isLocked(jsa) then
+                    mob:setLocalVar(string.format('sp_%u', jsa), os.time())
+                    debug(string.format('%s using %d', mob:getName(), jsa))
+                    mob:useMobAbility(jsa)
+                end
+            end
+        elseif #xi.av.sps ~= 0 then
+            local sp = xi.av.sps[math.random(1, #xi.av.sps)]
+            mob:setLocalVar(string.format('sp_%u', sp), os.time())
+            debug(string.format('%s using %d', mob:getName(), sp))
+            mob:useMobAbility(sp)
+        end
+
+        -- TODO: minimum should be max of 2hr combos length
+        xi.av.nextsp = now + math.random(45, 90)
+    end
+end
+
+entity.onMobInitialize = function(mob)
+    mob:setMobMod(xi.mobMod.ABILITY_RESPONSE, 1)
+
+    --[[
```

Comment:
Why are these commented out?

Reply:
I don't know anything about balance or gameplay and all of this is taken wholesale from Eden. I don't think we even have the same immunity system? So I opted to leave them commented for future reference